### PR TITLE
When we customize property serialization, also customize deserialization

### PR DESCRIPTION
**Pull request by havocp**
TLDR the client in examples/plotting/server/glyphs.py ended up with `inner_radius = { value: 10 }` instead of `inner_radius = { value: 10, units: 'screen' }` and now it keeps the units.

This fixes a problem with examples/plotting/server/glyphs.py where
we'd have AnnularWedge.inner_radius as a single value, we'd send it
to the client as a dict, at some point the client sent it back to us
as a dict, we then stripped the units out and emitted a ModelChangedEvent
as a dict sans units (because ModelChangedEvent contains the new regular
value, not the new serializable_value()). This new units-less dict was
then sent to the client causing the client to use default units instead
of the originally-provided units.

To fix this we need the serializable_value() in ModelChangedEvent.

We also need to be able to get a change event back from the client
without "forgetting" the dict-ness of our local attribute.
"dict-ness" means that we remember whether you set `foo = 42` or
`foo = { 'value' : 42 }`.

So we move the logic that does from_json and then sets the prop
from Document into a HasProps.set_from_json which in turn delegates
to Property, so that DataSpecProperty can process the JSON and
decide to preserve its dict-ness in the JSON case but not in the
regular `__set__` case.

### CODE REVIEW

File: bokeh/properties.py [hunk 1]
```diff
@@ -702,6 +708,27 @@ def __setattr__(self, name, value):
             raise AttributeError("unexpected attribute '%s' to %s, %s attributes are %s" %
                 (name, self.__class__.__name__, text, nice_join(matches)))
 
+    def set_from_json(self, name, json, models=None):
+        """Sets a property of the object using JSON and a dictionary from model ids to model instances.
+        The model instances are necessary if the JSON contains references to models.
+        """
+        if name in self.properties():
```
Comment (bryevdv) [inline bokeh/properties.py]: This `name` shadows the function parameter, is that intended?

**Comment by bryevdv** [comment]
tests look good, confirmed `glyphs.py` functioning properly again. LGTM

**Comment by bryevdv** [comment]
@mattpap any thoughts on this one?
